### ISSUE TITLE
Data leakage in the M4 experiments

### ISSUE THREAD

**Issue by linfeng-du** [thuml/Time-Series-Library#207]
Since `flag=val` and `flag=test` trigger the same logic in `Dataset_M4`, which is to load the test portion of the dataset. The `vali` function in `Exp_Short_Term_Forecast` is essentially the same as `test`, which leads to data leakage during training. Please correct me if I'm wrong.

A more proper way would be to validate using another sampled window from the training series.

**Comment by wuhaixu2016** [comment]
Hi, thanks for this question.
- We write this part following N-BEATS: https://github.com/ServiceNow/N-BEATS
- Validation set is proposed for model and hyperparameter selection. If you select the hyper-parameter or model based on your experience or other datasets, ranther than the validation loss, there are no data leaky.

**Comment by linfeng-du** [comment]
You are right but here it is using validation loss for early stopping which is also a part of model selection. Essentially the logic is to use the test data for early stopping here. The only difference is that when testing each sample is fed into TimesNet independently while during validation it's batched, hence the detected periodicities would be slightly different.

![image](https://github.com/thuml/Time-Series-Library/assets/34938020/473bbf83-6400-46c8-96bb-83022f921f87)
![image](https://github.com/thuml/Time-Series-Library/assets/34938020/f09f3776-76e5-418a-a353-6d18cd7ab470)
